tests/ui/unnecessary_option_map_or_else.rs

```diff
@@ -0,0 +1,54 @@
+#![warn(clippy::unnecessary_option_map_or_else)]
+#![allow(
+    clippy::let_and_return,
+    clippy::let_unit_value,
+    clippy::unnecessary_lazy_evaluations,
+    clippy::unnecessary_literal_unwrap
+)]
+
+fn main() {
+    // Expected errors
+    // Basic scenario
+    let option = Some(());
+    option.map_or_else(|| (), |x| x); //~ ERROR: unused "map closure" when calling
+
+    // Type ascription
+    let option = Some(());
+    option.map_or_else(|| (), |x: ()| x); //~ ERROR: unused "map closure" when calling
+
+    // Auto-deref
+    let string = String::new();
+    let option = Some(&string);
+    let _: &str = option.map_or_else(|| &string, |x| x); //~ ERROR: unused "map closure" when calling
+
+    // Temporary variable
+    let option = Some(());
+    option.map_or_else(
+        //~^ ERROR: unused "map closure" when calling
+        || (),
+        |x| {
+            let tmp = x;
+            tmp
+        },
+    );
+
+    // Correct usages
+    let option = Some(());
+    option.map_or_else(|| (), |_| ());
+
+    let option = Some(());
+    option.map_or_else(|| (), |_: ()| ());
+
+    let string = String::new();
+    let option = Some(&string);
+    let _: &str = option.map_or_else(|| &string, |_| &string);
+
+    let option = Some(());
+    option.map_or_else(
+        || (),
+        |_| {
+            let tmp = ();
+            tmp
+        },
+    );
+}
```

Comment:
I'd like to see a test case with `std::convert::identity`. And one where the closure is put in a `let` binding before the `map_or_else` (we can get that with `expr_or_init`).